README.txt

Lockette for Bukkit, written by Acru Jovian.
Distributed under the The Non-Profit Open Software License version 3.0 (NPOSL-3.0)
http://www.opensource.org/licenses/NOSL3.0

This project needs the following Libraries: (Version used in brackets)

* Bukkit (1.2.5-R5.0)
* PluginCore (1.3.5)

Included is an Ant build file to package into a single jar library.



changelog.txt
* Unreleased changes:

?


* Active release build change log is here:

http://dev.bukkit.org/server-mods/lockette/pages/lockette-change-log/


* But here is a copy/paste:

Version 1.7.3:

    Corrected an undefined method error.
    Added Norwegian language translation. 

Version 1.7.2:

    Added a configuration option to disable specific linked plugins, in case the api changes.
    Added a check for hoeing grass with attached trap doors.
    Re-implemented debug code as a configuration option, for test servers. 

Version 1.7.1: [ElgarL]

    Removed debug code which could be exploited on off-line servers. 

Version 1.7: [ElgarL]

    Allowed pluginCore to recognize/handle non Towny worlds.
    Update plugin dependencies to latest versions. 

Version 1.6:

    Lowered earliest supported build version to 1846 and made manditory.
    Fixed trees breaking signs off chests, this time for sure.
    Fixed sticky pistons pulling protected pullable blocks.
    Added Spanish language translation.
    Added protection of brewing stands. (Finally...)
    Added a custom list of protected blocks, set by config.
        Defaults to the three blocks made with diamonds.
        There may be some glitches, so be sure to test. 
    Core: Fixed native groups support for bPermissions.
    Core: Added LWC support for zones. 

Version 1.5

    Explosions are no longer canceled, but the list of blocks destroyed is pruned.
    Added check to prevent trees from growing into protected blocks.
    Fixed for the fence gate exploit.
    Added some additional translations.
    Added a command alias '/lock'.
    Built in support for color tags, when creating protection signs.
        Color tags are also allowed in the strings file. 
    Bug fix for piston extension event.
    Added a configuration option for operator breaking.
    Changed the format of the log messages, for a cleaner look.
    Core: Factions support for groups. (And fixed it again.)
    Core: SimpleClans support for groups.
    Core: MCMmo support for groups.
    Core: Superperms, along with bPermissions support.
    Core: Redesigned how other plugins are linked.
    Core: iConomy support. (Unused in Lockette 1.5 so far...)
    And more that I have probably forgotten... 

Version 1.4.4

    Added a feature to fix badly rotated chests when upgrading from pre-1.8 to 1.8+
        Right-clicking the side of an older chest will cause the chest to face that side, works once only.
        Added a configuration option 'enable-chest-rotation' to have this always work, defaults to false. 
    Core: Towny: Fixed never being allowed to protect chests in a town. 

Version 1.4.3

    Fixed an exception in the previous version, thrown when PermissionsBukkit/Towny is not available. 

Version 1.4.2

    Core: No longer attempts to force-enable PermissionsBukkit/PermissionsEx/Towny, due to a bug in Bukkit.
    Core: Improved Towny Town/Nation check for when a Resident/Town is not registered. 

Version 1.4.1

    Experimental Towny support, though not tested. I need someone to help with testing!
        Core: Towny group support for towns/nations.
        New permission lockette.towny.wilds for protecting in the wilds. 
    Fixed quick-protect throwing an exception when another plugin removed the placed sign.
    Core: New support of build zones.
    Core: Improved handling when using multiple permission/group plugins.
    Core: Added PermissionsBukkit group support. 

Version 1.4

    You can now protect a chest/furnace/dispenser instantly by right clicking it while holding a sign!
        A second sign will be a [More Users] sign with [Everyone] as a user.
        This new method can be toggled with the enable-quick-protect configuration option. 
    New command: /lockette fix, to toggle an automatic door that has gotten out of sync. Look at it when typing the command.
    Full trap door support.
    Full piston protection for doors and trap doors.
    Language file configuration option.
    Preliminary support for MC 1.8 fence gates, may or may not work as expected.
    Core: Worked around a change in GroupManager API, in the Essentials Phoenix build.
    Added missing sound effects for doors, supported in builds 851+.
    Fixed redstone still being disabled if the owner line is [Everyone].
    Applied a fix for when another plugin was triggering the conflicting door message, when there was no conflict.
    Improved security in a number of obscure cases.
    More public functions for plugin interoperability. 

Version 1.3.8

    Reversed the change related to MinecartManiaAdminControls, but said plugin is updated too now.
        This fixes a security issue. 

Version 1.3.7

    Worked around a change in MinecartManiaAdminControls 1.1.0 breaking door signs.
    Changed the default value for enable-permissions to false.
        Too many people had Permissions/GroupManager installed but didn't know how to use.
        This will not effect existing configuration files. 

Version 1.3.6

    Fixed a config file option forcing permissions on if there are no permission plugins.
    Added an internal warning about builds 685-703. 


src/org/yi/acru/bukkit/Lockette/LocketteDoorCloser.java
//
// This file is a component of Lockette for Bukkit, and was written by Acru Jovian.
// Distributed under the The Non-Profit Open Software License version 3.0 (NPOSL-3.0)
// http://www.opensource.org/licenses/NOSL3.0
//


package org.yi.acru.bukkit.Lockette;


// Imports.
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;

import org.bukkit.World;
import org.bukkit.block.Block;



public class LocketteDoorCloser implements Runnable{
	private static Lockette					plugin;
	private static int						doorTask = -1;
	private final PriorityQueue<closeTask>	closeTaskList = new PriorityQueue<closeTask>();
	
	
	
	public LocketteDoorCloser(Lockette instance){
		plugin = instance;
	}
	
	
	protected boolean start(){
		if(doorTask != -1) return(false);
		
		doorTask = plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, this, 100, 10);
		if(doorTask == -1) return(true);
		
		return(false);
	}
	
	
	protected boolean stop(){
		if(doorTask == -1) return(false);
		
		plugin.getServer().getScheduler().cancelTask(doorTask);
		doorTask = -1;
		cleanup();
		
		return(false);
	}
	
	
	protected void cleanup(){
		closeTask		door;
		
		while(!closeTaskList.isEmpty()){
			door = closeTaskList.poll();
			if(door == null) break;
			
			close(door);
		}
	}
	
	
	public void run(){
		if(closeTaskList.isEmpty()) return;
		
		Date		time = new Date();
		closeTask	door;
		
		while(time.after(closeTaskList.peek().time)){
			door = closeTaskList.poll();
			if(door == null) break;
			
			close(door);
			if(closeTaskList.isEmpty()) break;
		}
	}
	
	
	private void close(closeTask door){
		//Lockette.log.info("Closing at " + door.time);
		Lockette.toggleHalfDoor(door.world.getBlockAt(door.x, door.y, door.z), door.effect);
	}
	
	
	// Assumes all blocks in the list are for the same world.
	public void add(List<Block> list, boolean auto, int delta){
		if(list == null) return;
		if(list.isEmpty()) return;
		
		Iterator<closeTask>	it;
		Iterator<Block>		itb;
		
		closeTask	task;
		Block		block;
		World		world = list.get(0).getWorld();
		
		// Check each item in the task list for duplicates.
		it = closeTaskList.iterator();
		while(it.hasNext()){
			task = it.next();
			
			// Ignore block lists for another world.
			if(!task.world.equals(world)) continue;
			
			// Compare to each of the blocks in the new list.
			itb = list.iterator();
			while(itb.hasNext()){
				block = itb.next();
				
				// Ignore block lists for another world.
				//if(!block.getWorld().equals(task.world)) break;
				
				if((block.getX() == task.x) && (block.getY() == task.y) && (block.getZ() == task.z)){
					// Duplicate found, remove it from both lists.  A door to be closed was closed manually.
					//Lockette.log.info("Duplicate");
					it.remove();
					itb.remove();
					break;
				}
			}
		}
		
		
		// If the list is empty or the new door blocks aren't automatic, we are done.
		if(!auto) return;
		if(list.isEmpty()) return;
		
		
		// Add the new blocks to the task list.
		//(((block.getX()&0x7)*4)+2)
		//Lockette.log.info("Adding " + list.size());
		
		Date		time = new Date();
		time.setTime(time.getTime() + (delta * 1000L));
		
		for(int x = 0; x < list.size(); ++x){
			closeTaskList.add(new closeTask(time, list.get(x), x == 0));
		}
	}
	
	
	protected class closeTask implements Comparable<closeTask>{
		Date		time;
		World		world;
		int			x, y, z;
		boolean		effect;
		
		
		public closeTask(Date taskTime, Block block, boolean taskEffect){
			time = taskTime;
			world = block.getWorld();
			x = block.getX();
			y = block.getY();
			z = block.getZ();
			effect = taskEffect;
		}
		
		
		public int compareTo(closeTask arg){
			return(time.compareTo(arg.time));
		}
	}
}



src/org/yi/acru/bukkit/Lockette/LocketteWorldListener.java
//
// This file is a component of Lockette for Bukkit, and was written by Acru Jovian.
// Distributed under the The Non-Profit Open Software License version 3.0 (NPOSL-3.0)
// http://www.opensource.org/licenses/NOSL3.0
//


package org.yi.acru.bukkit.Lockette;

//Imports.
import java.util.List;

import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockState;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.plugin.PluginManager;

import org.bukkit.event.world.StructureGrowEvent;



public class LocketteWorldListener implements Listener{
	private static Lockette		plugin;
	
	
	public LocketteWorldListener(Lockette instance){
		plugin = instance;
	}
	
	
	protected void registerEvents(){
		PluginManager	pm = plugin.getServer().getPluginManager();
		
		pm.registerEvents(this, plugin);
	}
	
	
	//********************************************************************************************************************
	// Start of event section
	
	
	@EventHandler(priority = EventPriority.HIGHEST)
	public void onStructureGrow(StructureGrowEvent event){
		if(event.isCancelled()) return;
		
		List<BlockState> blockList = event.getBlocks();
		int			x, count = blockList.size();
		Block		block;
		
		/*
		// Structure debug routine, all I touch turns to glass.
		
		if(count >= 0){
			//Lockette.log.info("[Lockette] All I touch turns to glass. (" + count + ")");
			plugin.getServer().broadcastMessage("[Lockette] All I touch turns to glass. (" + count + ")");
			
			for(x = 0; x < count; ++x){
				block = blockList.get(x).getBlock();
				block.setTypeId(20);
			}
			
			event.setCancelled(true);
			return;
		}
		*/
		
		// Check the block list for any protected blocks, and cancel the event if any are found.
		
		for(x = 0; x < count; ++x){
			block = blockList.get(x).getBlock();
			
			if(Lockette.isProtected(block)){
				event.setCancelled(true);
				return;
			}
			
			if(Lockette.explosionProtectionAll){
				if((block.getTypeId() == Material.CHEST.getId()) || (block.getTypeId() == Material.DISPENSER.getId()) ||
						(block.getTypeId() == Material.FURNACE.getId()) || (block.getTypeId() == Material.BURNING_FURNACE.getId()) ||
						(block.getTypeId() == Material.BREWING_STAND.getId())){
					event.setCancelled(true);
					return;
				}
			}
		}
	}
}



src/org/yi/acru/bukkit/Lockette/LocketteEntityListener.java
//
// This file is a component of Lockette for Bukkit, and was written by Acru Jovian.
// Distributed under the The Non-Profit Open Software License version 3.0 (NPOSL-3.0)
// http://www.opensource.org/licenses/NOSL3.0
//


package org.yi.acru.bukkit.Lockette;

// Imports.
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.plugin.PluginManager;

import org.bukkit.event.entity.EntityExplodeEvent;



public class LocketteEntityListener implements Listener{
	private static Lockette		plugin;
	
	
	public LocketteEntityListener(Lockette instance){
		plugin = instance;
	}
	
	
	protected void registerEvents(){
		PluginManager	pm = plugin.getServer().getPluginManager();
		
		pm.registerEvents(this, plugin);
	}
	
	
	//********************************************************************************************************************
	// Start of event section
	
	
	@EventHandler(priority = EventPriority.HIGHEST)
	public void onEntityExplode(EntityExplodeEvent event){
		if(event.isCancelled()) return;
		
		//List<Block> blockList = event.blockList();
		int			x;//, count = event.blockList().size();
		Block		block;
		
		
		// Check the block list for any protected blocks, and cancel the event if any are found.
		
		for(x = 0; x < event.blockList().size(); ++x){
			block = event.blockList().get(x);
			
			if(Lockette.isProtected(block)){
				//event.setCancelled(true);
				//return;
				event.blockList().remove(x);
				--x;
				//--count;
				continue;
			}
			
			if(Lockette.explosionProtectionAll){
				if((block.getTypeId() == Material.CHEST.getId()) || (block.getTypeId() == Material.DISPENSER.getId()) ||
						(block.getTypeId() == Material.FURNACE.getId()) || (block.getTypeId() == Material.BURNING_FURNACE.getId()) ||
						(block.getTypeId() == Material.BREWING_STAND.getId())){
					//event.setCancelled(true);
					//return;
					event.blockList().remove(x);
					--x;
					//--count;
					continue;
				}
			}
		}
	}
}



src/org/yi/acru/bukkit/Lockette/LockettePlayerListener.java
//
// This file is a component of Lockette for Bukkit, and was written by Acru Jovian.
// Distributed under the The Non-Profit Open Software License version 3.0 (NPOSL-3.0)
// http://www.opensource.org/licenses/NOSL3.0
//


package org.yi.acru.bukkit.Lockette;

// Imports.
import java.util.List;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Sign;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.Event.Result;
import org.bukkit.event.block.Action;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.PluginManager;

import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerQuitEvent;



public class LockettePlayerListener implements Listener{
	private static Lockette		plugin;
	
	final static int		materialFenceGate = 107;
	
	
	public LockettePlayerListener(Lockette instance){
		plugin = instance;
	}
	
	
	protected void registerEvents(){
		PluginManager	pm = plugin.getServer().getPluginManager();

		pm.registerEvents(this, plugin);
	}
	
	
	//********************************************************************************************************************
	// Start of event section
	
	
	@EventHandler(priority = EventPriority.NORMAL)
	public void onPlayerCommandPreprocess(PlayerCommandPreprocessEvent event){
		String[]	command = event.getMessage().split(" ", 3);
		
		if(command.length < 1) return;
		if(!(command[0].equalsIgnoreCase("/lockette") || command[0].equalsIgnoreCase("/lock"))) return;
		event.setCancelled(true);
		
		Player		player = event.getPlayer();
		
		
		// Reload config files, for admins only.
		if(command.length == 2){
			if(command[1].equalsIgnoreCase("reload")){
				if(!plugin.hasPermission(player.getWorld(), player, "lockette.admin.reload")) return;
				
				plugin.loadProperties(true);
				
				plugin.localizedMessage(player, Lockette.broadcastReloadTarget, "msg-admin-reload");
				return;
			}
			
			if(command[1].equalsIgnoreCase("version")){
				player.sendMessage(ChatColor.RED + "Lockette version " + plugin.getDescription().getVersion() + " loaded.  (Core: " + Lockette.getCoreVersion() + ")");
				return;
			}
			
			if(command[1].equalsIgnoreCase("fix")){
				if(fixDoor(player)){
					plugin.localizedMessage(player, null, "msg-error-fix");
				}
				return;
			}
		}
		
		
		// Edit sign text.
		if((command.length == 2) || (command.length == 3)){
			if(command[1].equals("1") || command[1].equals("2") || command[1].equals("3") || command[1].equals("4")){
				Block		block = plugin.playerList.get(player.getName());
				//boolean		error = false;
				
				// Check if the selected block is a valid sign.
				
				if(block == null){plugin.localizedMessage(player, null, "msg-error-edit"); return;}
				else if(block.getTypeId() != Material.WALL_SIGN.getId()){plugin.localizedMessage(player, null, "msg-error-edit"); return;}
				

				Sign		sign = (Sign) block.getState();
				Sign		owner = sign;
				String		text = sign.getLine(0).replaceAll("(?i)\u00A7[0-F]", "").toLowerCase();
				boolean		privateSign;
				
				// Check if it is our sign that is selected.
				
				if(text.equals("[private]") || text.equalsIgnoreCase(Lockette.altPrivate)) privateSign = true;
				else if(text.equals("[more users]") || text.equalsIgnoreCase(Lockette.altMoreUsers)){
					privateSign = false;
					
					Block		checkBlock = Lockette.getSignAttachedBlock(block);
					if(checkBlock == null){plugin.localizedMessage(player, null, "msg-error-edit"); return;}
					
					Block		signBlock = Lockette.findBlockOwner(checkBlock);
					if(signBlock == null){plugin.localizedMessage(player, null, "msg-error-edit"); return;}
					
					owner = (Sign) signBlock.getState();
				}
				else{plugin.localizedMessage(player, null, "msg-error-edit"); return;}
				
				
				int			length = player.getName().length();
				
				if(length > 15) length = 15;

				// Check owner.
				if(owner.getLine(1).replaceAll("(?i)\u00A7[0-F]", "").equals(player.getName().substring(0, length)) || Lockette.debugMode){
					int			line = Integer.parseInt(command[1]) - 1;
					
					// Disallow editing [Private] line 1/2 here.
					if(!Lockette.debugMode){
						if(line <= 0) return;
						else if(line <= 1) if(privateSign) return;
					}
					
					
					if(command.length == 3){
						length = command[2].length();
						
						if(length > 15) length = 15;
						if(Lockette.colorTags) sign.setLine(line, command[2].substring(0, length).replaceAll("&([0-9A-Fa-f])", "\u00A7$1"));
						else sign.setLine(line, command[2].substring(0, length));
					}
					else sign.setLine(line, "");
					sign.update();
					
					plugin.localizedMessage(player, null, "msg-owner-edit");
					return;
				}
				else{plugin.localizedMessage(player, null, "msg-error-edit"); return;}
			}
		}
		
		
		// If none of the above, print out the help text.
		// Commands:
		// reload
		// 2-4 <text> - sign editing
		// link - linking?
		// set <value> <string> - config?

		plugin.localizedMessage(player, null, "msg-help-command1");
		plugin.localizedMessage(player, null, "msg-help-command2");
		plugin.localizedMessage(player, null, "msg-help-command3");
		plugin.localizedMessage(player, null, "msg-help-command4");
		plugin.localizedMessage(player, null, "msg-help-command5");
		plugin.localizedMessage(player, null, "msg-help-command6");
		plugin.localizedMessage(player, null, "msg-help-command7");
		plugin.localizedMessage(player, null, "msg-help-command8");
		plugin.localizedMessage(player, null, "msg-help-command9");
	}
	
	
	@EventHandler(priority = EventPriority.HIGHEST)
	public void onPlayerInteract(PlayerInteractEvent event){
		if(!event.hasBlock()) return;
		
		Action		action = event.getAction();
		Player		player = event.getPlayer();
		Block		block = event.getClickedBlock();
		int			type = block.getTypeId();
		BlockFace	face = event.getBlockFace();
		
		
		if(action == Action.RIGHT_CLICK_BLOCK){
			if(Lockette.protectTrapDoors) if(type == Material.TRAP_DOOR.getId()){
				if(interactDoor(block, player)) return;
				
				event.setUseInteractedBlock(Result.DENY);
				event.setUseItemInHand(Result.DENY);
				return;
			}
			
			if(Lockette.protectDoors) if((type == Material.WOODEN_DOOR.getId()) || (type == Material.IRON_DOOR_BLOCK.getId()) || (type == materialFenceGate)){
				if(interactDoor(block, player)) return;
				
				event.setUseInteractedBlock(Result.DENY);
				event.setUseItemInHand(Result.DENY);
				return;
			}
			
			if(type == Material.WALL_SIGN.getId()){
				interactSign(block, player);
				return;
			}
			
			if(type == Material.CHEST.getId()){
				// Try at making a 1.7->1.8 chest fixer.
				Lockette.rotateChestOrientation(block, face);
			}
			
			if((type == Material.CHEST.getId()) || (type == Material.DISPENSER.getId()) ||
					(type == Material.FURNACE.getId()) || (type == Material.BURNING_FURNACE.getId()) ||
					(type == Material.BREWING_STAND.getId()) || Lockette.isInList(type, Lockette.customBlockList)){
				
				// Trying something out....
				if(Lockette.directPlacement) if(event.hasItem()) if((face != BlockFace.UP) && (face != BlockFace.DOWN)){
					ItemStack		item = event.getItem();
					
					if(item.getTypeId() == Material.SIGN.getId()){
						Block		checkBlock = block.getRelative(face);
						
						type = checkBlock.getTypeId();
						
						if(type == Material.AIR.getId()){
							boolean		place = false;
							
							if(Lockette.isProtected(block)){
								// Add a users sign only if owner.
								if(Lockette.isOwner(block, player.getName())) place = true; 
							}
							else place = true;
							//if(Lockette.altPrivate == null){}//if(Lockette.altMoreUsers == null){}
							if(place){
								//player.sendMessage(ChatColor.RED + "Lockette: Using a sign on a container");
								
								event.setUseItemInHand(Result.ALLOW); //? seems to work in 568
								event.setUseInteractedBlock(Result.DENY);
								return;
							}
						}
					}
				}
				
				
				if(interactContainer(block, player)) return;
				
				event.setUseInteractedBlock(Result.DENY);
				event.setUseItemInHand(Result.DENY);
				return;
			}
			
			if(type == Material.DIRT.getId()) if(event.hasItem()){
				ItemStack		item = event.getItem();
				
				type = item.getTypeId();
				
				if((type == Material.DIAMOND_HOE.getId()) || (type == Material.GOLD_HOE.getId()) || (type == Material.IRON_HOE.getId()) || 
						(type == Material.STONE_HOE.getId()) || (type == Material.WOOD_HOE.getId())){
					Block		checkBlock = block.getRelative(BlockFace.UP);
					
					type = checkBlock.getTypeId();
					
					if((type == Material.WOODEN_DOOR.getId()) || (type == Material.IRON_DOOR_BLOCK.getId()) || (type == materialFenceGate)){
						event.setUseInteractedBlock(Result.DENY);
						return;
					}
					
					if(hasAttachedTrapDoor(block)){
						event.setUseInteractedBlock(Result.DENY);
						return;
					}
				}
			}
		}
		else if(action == Action.LEFT_CLICK_BLOCK){
			if(Lockette.protectTrapDoors) if(type == Material.TRAP_DOOR.getId()){
				if(interactDoor(block, player)) return;
				
				event.setUseInteractedBlock(Result.DENY);
				event.setUseItemInHand(Result.DENY);
				return;
			}
			
			if(Lockette.protectDoors) if((type == Material.WOODEN_DOOR.getId()) || (type == Material.IRON_DOOR_BLOCK.getId()) || (type == materialFenceGate)){
				if(interactDoor(block, player)) return;
				
				event.setUseInteractedBlock(Result.DENY);
				event.setUseItemInHand(Result.DENY);
				return;
			}
		}
	}
	
	
	@EventHandler(priority = EventPriority.NORMAL)
	public void onPlayerQuit(PlayerQuitEvent event){
		Player		player = event.getPlayer();
		
		// Player left, so forget about them.
		plugin.playerList.remove(player.getName());
	}
	
	
	//********************************************************************************************************************
	// Start of interact section
	
	
	// Returns true if it should be allowed, false if it should be canceled.
	private static boolean interactDoor(Block block, Player player){
		Block		signBlock = Lockette.findBlockOwner(block);
		
		if(signBlock == null) return(true);
		
		boolean		wooden = (block.getTypeId() == Material.WOODEN_DOOR.getId());
		boolean		trap = false;
		
		if(block.getTypeId() == materialFenceGate) wooden = true;
		if(Lockette.protectTrapDoors) if(block.getTypeId() == Material.TRAP_DOOR.getId()){
			wooden = true;
			trap = true;
		}
		
		
		// Someone touched an owned door, lets see if they are allowed.
		
		boolean		allow = false;
		
		if(canInteract(block, signBlock, player, true)) allow = true;
		/*
		// Fee stuff...
		if(!allow){
			// Check if we can pay a fee to activate.
			int fee = getSignOption(signBlock, "fee", Lockette.altFee, 0);
			
			if(signBlock.equals(plugin.playerList.get(player.getName()))){
				if(fee == 0){
					player.sendMessage("unable to pay fee");
				}
				else{
					player.sendMessage("Fee of " +plugin.economyFormat(fee)+" paid (fake)");
					plugin.playerList.put(player.getName(), block);
					allow = true;
				}
				/
				if(fee != 0){
					Sign		sign = (Sign) signBlock.getState();
					String		text = sign.getLine(1).replaceAll("(?i)\u00A7[0-F]", "");
					
					if(plugin.economyTransfer(player.getName(), text, fee)){
						allow = true;
						plugin.playerList.put(player.getName(), block);
					}
					else{}
				}/
			}
			else if(fee != 0){
				player.sendMessage("first touch sign to pay fee.");
			}
		}
		*/
		if(allow){
			List<Block> list = Lockette.toggleDoors(block, Lockette.getSignAttachedBlock(signBlock), wooden, trap);
			
			int delta = Lockette.getSignOption(signBlock, "timer", Lockette.altTimer, Lockette.defaultDoorTimer);
			
			plugin.doorCloser.add(list, delta != 0, delta);
			return(true);
		}
		
		
		// Don't have permission.
		
		//event.setCancelled(true);
		
		//if(Lockette.oldListener){
		//	if(wooden) Lockette.toggleSingleDoor(block);
		//}
		
		// Report only once, unless a different block is clicked.
		if(block.equals(plugin.playerList.get(player.getName()))) return(false);
		plugin.playerList.put(player.getName(), block);
		plugin.localizedMessage(player, null, "msg-user-denied-door");
		return(false);
	}
	
	
	private static void interactSign(Block block, Player player){
		Sign		sign = (Sign) block.getState();
		String		text = sign.getLine(0).replaceAll("(?i)\u00A7[0-F]", "").toLowerCase();
		Block		signBlock = block;
		
		
		// Check if it is our sign that was clicked.
		
		if(text.equals("[private]") || text.equalsIgnoreCase(Lockette.altPrivate)){}
		else if(text.equals("[more users]") || text.equalsIgnoreCase(Lockette.altMoreUsers)){
			Block		checkBlock = Lockette.getSignAttachedBlock(block);
			if(checkBlock == null) return;
			
			signBlock = Lockette.findBlockOwner(checkBlock);
			if(signBlock == null) return;
			
			sign = (Sign) signBlock.getState();
		}
		else return;
		
		int			length = player.getName().length();
		
		if(length > 15) length = 15;
		
		// Check owner.
		if(sign.getLine(1).replaceAll("(?i)\u00A7[0-F]", "").equals(player.getName().substring(0, length)) || Lockette.debugMode){
			if(!block.equals(plugin.playerList.get(player.getName()))){
				// Associate the user with the owned sign.
				plugin.playerList.put(player.getName(), block);
				plugin.localizedMessage(player, null, "msg-help-select");
			}
		}
		else{/*
			int fee = getSignOption(signBlock, "fee", Lockette.altFee, 0);
			
			if(fee != 0){
				if(!signBlock.equals(plugin.playerList.get(player.getName()))){
					// First half of fee approval.
					plugin.playerList.put(player.getName(), signBlock);
					plugin.localizedMessage(player, null, "msg-user-touch-fee", sign.getLine(1), plugin.economyFormat(fee));
				}
			}
			else{*/
				if(!block.equals(plugin.playerList.get(player.getName()))){
					// Only print this message once as well.
					plugin.playerList.put(player.getName(), block);
					plugin.localizedMessage(player, null, "msg-user-touch-owned", sign.getLine(1));
				}
			//}
		}
	}
	
	
	// Returns true if it should be allowed, false if it should be canceled.
	private static boolean interactContainer(Block block, Player player){
		Block		signBlock = Lockette.findBlockOwner(block);
		
		if(signBlock == null) return(true);
		
		
		// Someone touched an owned container, lets see if they are allowed.
		
		if(canInteract(block, signBlock, player, false)) return(true);
		
		
		// Don't have permission.
		
		// Report only once, unless a different block is clicked.
		if(block.equals(plugin.playerList.get(player.getName()))) return(false);
		plugin.playerList.put(player.getName(), block);
		plugin.localizedMessage(player, null, "msg-user-denied");
		return(false);
	}
	
	
	// Block is the container or door, signBlock is the owning [Private] sign.
	// Returns true if it should be allowed, false if it should be canceled.
	private static boolean canInteract(Block block, Block signBlock, Player player, boolean isDoor){
		// Check if the block is owned first.
		
		// Moved to outer..
		
		
		// Lets see if the player is allowed to touch...
		
		Sign		sign = (Sign) signBlock.getState();
		int			length = player.getName().length();
		String		line;
		
		if(length > 15) length = 15;
		
		
		// Check owner.
		
		line = sign.getLine(1).replaceAll("(?i)\u00A7[0-F]", "");
		
		if(line.equals(player.getName().substring(0, length))) return(true);
		if(plugin.inGroup(block.getWorld(), player, line)) return(true);
		
		
		// Check main two users.
		
		int			y;
		
		for(y = 2; y <= 3; ++y) if(!sign.getLine(y).isEmpty()){
			line = sign.getLine(y).replaceAll("(?i)\u00A7[0-F]", "");
			
			if(plugin.inGroup(block.getWorld(), player, line)) return(true);
			if(line.equalsIgnoreCase(player.getName().substring(0, length))) return(true);
		}
		
		
		// Check for more users.
		
		List<Block>	list = Lockette.findBlockUsers(block, signBlock);
		int			x, count = list.size();
		Sign		sign2;
		
		for(x = 0; x < count; ++x){
			sign2 = (Sign) list.get(x).getState();
			
			for(y = 1; y <= 3; ++y) if(!sign2.getLine(y).isEmpty()){
				line = sign2.getLine(y).replaceAll("(?i)\u00A7[0-F]", "");
				
				if(plugin.inGroup(block.getWorld(), player, line)) return(true);
				if(line.equalsIgnoreCase(player.getName().substring(0, length))) return(true);
			}
		}
		
		
		// Check admin list last.
		
		boolean		snoop = false;
		
		if(isDoor){
			if(Lockette.adminBypass){
				if(plugin.hasPermission(block.getWorld(), player, "lockette.admin.bypass")) snoop = true;
				
				if(snoop){
					Lockette.log.info("[" + plugin.getDescription().getName() + "] (Admin) " + player.getName() + " has bypassed a door owned by " + sign.getLine(1));
					
					plugin.localizedMessage(player, null, "msg-admin-bypass", sign.getLine(1));
					return(true);
				}
			}
		}
		else if(Lockette.adminSnoop){
			if(plugin.hasPermission(block.getWorld(), player, "lockette.admin.snoop")) snoop = true;
			
			if(snoop){
				Lockette.log.info("[" + plugin.getDescription().getName() + "] (Admin) " + player.getName() + " has snooped around in a container owned by " + sign.getLine(1) + "!");
				
				plugin.localizedMessage(player, Lockette.broadcastSnoopTarget, "msg-admin-snoop", sign.getLine(1));
				return(true);
			}
		}
		
		
		// Don't have permission.
		
		return(false);
	}
	
	
	//********************************************************************************************************************
	// Start of utility section
	
	
	// Returns true if a door wasn't changed.
	private static boolean fixDoor(Player player){
		Block		block = player.getTargetBlock(null, 10);
		int			type = block.getTypeId();
		boolean		doCheck = false;
		
		
		// Check if the block being looked at is a door block.
		
		if(Lockette.protectTrapDoors){
			if(type == Material.TRAP_DOOR.getId()) doCheck = true;
		}
		
		if(Lockette.protectDoors){
			if((type == Material.WOODEN_DOOR.getId()) || (type == Material.IRON_DOOR_BLOCK.getId()) || (type == materialFenceGate)) doCheck = true;
		}
		
		if(!doCheck) return(true);
		
		
		Block		signBlock = Lockette.findBlockOwner(block);
		
		if(signBlock == null) return(true);
		
		Sign		sign = (Sign) signBlock.getState();
		int			length = player.getName().length();
		
		if(length > 15) length = 15;
		
		// Check owner only.
		if(sign.getLine(1).replaceAll("(?i)\u00A7[0-F]", "").equals(player.getName().substring(0, length))){
			Lockette.toggleSingleDoor(block);
			return(false);
		}
		
		return(true);
	}
	
	
	public static boolean hasAttachedTrapDoor(Block block){
		Block		checkBlock;
		int			type;
		int			face;
		
		checkBlock = block.getRelative(BlockFace.NORTH);
		type = checkBlock.getTypeId();
		if(type == Material.TRAP_DOOR.getId()){
			face = checkBlock.getData() & 0x3;
			if(face == 2) return(true);
		}
		
		checkBlock = block.getRelative(BlockFace.EAST);
		type = checkBlock.getTypeId();
		if(type == Material.TRAP_DOOR.getId()){
			face = checkBlock.getData() & 0x3;
			if(face == 0) return(true);
		}
		
		checkBlock = block.getRelative(BlockFace.SOUTH);
		type = checkBlock.getTypeId();
		if(type == Material.TRAP_DOOR.getId()){
			face = checkBlock.getData() & 0x3;
			if(face == 3) return(true);
		}
		
		checkBlock = block.getRelative(BlockFace.WEST);
		type = checkBlock.getTypeId();
		if(type == Material.TRAP_DOOR.getId()){
			face = checkBlock.getData() & 0x3;
			if(face == 1) return(true);
		}
		
		return(false);
	}
}

